### PR TITLE
[Clang] Add [[clang::no_specializations]]

### PR DESCRIPTION
**Pull request by philnik777**
This can be used to inform users when a template should not be specialized. For example, this is the case for the standard type traits (except for `common_type` and `common_reference`, which have more complicated rules).

### CODE REVIEW

**Comment by MitalAshok** [comment]
A list of places this could be used:

https://eel.is/c++draft/cmp.result#1
https://eel.is/c++draft/meta.rqmts#4
https://eel.is/c++draft/execpol.type#3
https://eel.is/c++draft/format.arg#2
https://eel.is/c++draft/range.adaptor.object#5
https://eel.is/c++draft/coro.generator.class#4
https://eel.is/c++draft/time.traits.is.clock#2
https://eel.is/c++draft/saferecl.rcu.base#2
https://eel.is/c++draft/saferecl.hp.base#2

Also consider adding an argument that lets you diagnose https://eel.is/c++draft/namespace.std#2.1

>  - the added declaration depends on at least one program-defined type,

To diagnose specializations for `std::pair<int, int>`, but allow `std::pair<MyClass, int>`.

File: clang/include/clang/Basic/AttrDocs.td [hunk 1]
```diff
@@ -975,6 +975,15 @@ Query for this feature with ``__has_attribute(diagnose_if)``.
   }];
 }
 
+def DiagnoseSpecializationsDocs : Documentation {
+  let Category = DocCatDecl;
+  let Content = [{
+``clang::diagnose_specializations`` can be appied to class templates which
+should not be specialized by users. This is primarily used to diagnose user
+specializations of standard library type traits.
```
Comment (MitalAshok) [inline clang/include/clang/Basic/AttrDocs.td:983]: "applied to class or variable templates which should not be specialized."

I think "by users" should either be omitted or replaced with "by others"/"elsewhere" here, since this could be used by non-standard-library writers too. It's fine in the next sentence though.

Comment (philnik777) [reply clang/include/clang/Basic/AttrDocs.td:983]: Are users of a non-standard-library not users?

Comment (MitalAshok) [reply clang/include/clang/Basic/AttrDocs.td]: Oh right, I think I misread it.

You still have the "appied"->"applied" typo btw.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 2]
```diff
@@ -5407,6 +5407,9 @@ def note_dependent_function_template_spec_discard_reason : Note<
   "candidate ignored: %select{not a function template|"
   "not a member of the enclosing %select{class template|"
   "namespace; did you mean to explicitly qualify the specialization?}1}0">;
+def warn_diag_specialization : Warning<
```
Comment (Sirraide) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:5410]: nit: `warn_diag_...` is a bit of a weird name for a diagnostic. Can we call this just `warn_invalid_specialization` or sth like that?

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 3]
```diff
@@ -5407,6 +5407,9 @@ def note_dependent_function_template_spec_discard_reason : Note<
...
+  "specializing a template which should not be specialized">,
```
Comment (Sirraide) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:5411]: I fee; like rephrasing it to something along the lines of ‘'X' should not be specialised’ would make it a bit clearer what the problem is.

‘Specialising a template which should not be specialised’ is probably a bit unclear to anyone who doesn’t already know about this rule, whereas e.g. ‘`std::something` should not be specialised’ imo makes it clearer what exactly the problem here is (i.e. that *particular* template should not be specialised).

**Review by Sirraide — COMMENTED** [review]
This is missing a release note, and imo rephrasing the diagnostic a bit would make sense, but the rest seems fine to me.

File: clang/test/SemaCXX/attr-diagnose-specializations.cpp [hunk 4]
```diff
@@ -0,0 +1,34 @@
+// RUN: %clang_cc1 %s -verify
+
+#if !__has_cpp_attribute(clang::diagnose_specializations)
+#  error
+#endif
+
+struct [[clang::diagnose_specializations]] S {}; // expected-warning {{'diagnose_specializations' attribute only applies to class templates}}
+
+template <class T, class U>
+struct [[clang::diagnose_specializations]] is_same {
```
Comment (ldionne) [inline clang/test/SemaCXX/attr-diagnose-specializations.cpp:10]: Would it make sense to allow passing a string (maybe optionally) to explain why the template shouldn't be specialized? Then from libc++ we would do:

```
#define _LIBCPP_NO_SPECIALIZATION [[_Clang::__diagnose_specializations__("Per <section> of the C++ Standard, users are not allowed to specialize class templates in namespace std")]]
```

Comment (philnik777) [reply clang/test/SemaCXX/attr-diagnose-specializations.cpp:10]: I guess we could add an optional string, but I'd actually keep the error message generic in libc++. IMO adding the stable name is more useful for us than for users. Maybe a bit stronger wording like `'X' cannot be specialized` would make sense.

File: clang/include/clang/Basic/DiagnosticSemaKinds.td [hunk 5]
```diff
@@ -5408,7 +5408,10 @@ def note_dependent_function_template_spec_discard_reason : Note<
   "not a member of the enclosing %select{class template|"
   "namespace; did you mean to explicitly qualify the specialization?}1}0">;
 def warn_invalid_specialization : Warning<
-  "%0 should not be specialized">,
+  "%0 cannot be specialized">,
```
Comment (MitalAshok) [inline clang/include/clang/Basic/DiagnosticSemaKinds.td:5411]:
```suggestion
  "%0 cannot be specialized%select{|: %2}1">,
```

And then change the diagnostics to `Diag(Loc, diag::warn_invalid_specialization) << VarOrClassTemplate << !Message.empty() << Message;`

File: clang/include/clang/Basic/Attr.td [hunk 6]
```diff
@@ -3327,6 +3330,15 @@ def DiagnoseIf : InheritableAttr {
   let Documentation = [DiagnoseIfDocs];
 }
 
+def DiagnoseSpecializations : InheritableAttr {
+  let Spellings = [Clang<"diagnose_specializations", /*AllowInC*/0>];
+  let Args = [StringArgument<"Message", 1>];
```
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:3335]: Do we want to allow the user to decide between warning vs error like we do for `diagnose_if`?

File: clang/include/clang/Basic/Attr.td [hunk 7]
```diff
@@ -3327,6 +3330,15 @@ def DiagnoseIf : InheritableAttr {
   let Documentation = [DiagnoseIfDocs];
 }
 
+def DiagnoseSpecializations : InheritableAttr {
+  let Spellings = [Clang<"diagnose_specializations", /*AllowInC*/0>];
```
Comment (AaronBallman) [inline clang/include/clang/Basic/Attr.td:3334]: Should this be named `diagnose_explicit_specializations` given: https://eel.is/c++draft/temp.spec#general-4.sentence-2

File: clang/include/clang/Basic/AttrDocs.td [hunk 8]
```diff
@@ -975,6 +975,15 @@ Query for this feature with ``__has_attribute(diagnose_if)``.
   }];
 }
 
+def DiagnoseSpecializationsDocs : Documentation {
+  let Category = DocCatDecl;
+  let Content = [{
+``[[clang::diagnose_specializations]]`` can be applied to class or variable
+templates which should not be specialized by users. This is primarily used to
```
Comment (AaronBallman) [inline clang/include/clang/Basic/AttrDocs.td:982]:
```suggestion
templates which should not be explicitly specialized by users. This is primarily used to
```

File: clang/test/SemaCXX/attr-diagnose-specializations.cpp [hunk 9]
```diff
@@ -0,0 +1,48 @@
+// RUN: %clang_cc1 %s -verify
+
+#if !__has_cpp_attribute(clang::diagnose_specializations)
+#  error
+#endif
+
+struct [[clang::diagnose_specializations]] S {}; // expected-warning {{'diagnose_specializations' attribute only applies to class templates}}
```
Comment (AaronBallman) [inline clang/test/SemaCXX/attr-diagnose-specializations.cpp:7]: doesn't it also apply to variable templates?

File: clang/include/clang/Basic/Attr.td (hunk 7 above)
Comment (philnik777) [reply clang/include/clang/Basic/Attr.td:3334]: Do explicit specializations include partial specializations?

File: clang/include/clang/Basic/Attr.td (hunk 6 above)
Comment (philnik777) [reply clang/include/clang/Basic/Attr.td:3335]: The idea of this attribute is that it says "if you specialize this anything could happen, so don't do it". I don't think it makes a ton of sense to give the choice of making this a warning by default on the attribute level.

**Review by cor3ntin — COMMENTED** [review]
I like the general idea 


- I would prefer this to be called `no_specializations` or `disallow_specializations`
- I would like to see more tests to what it can mor cannot be applied to (for example an explicit specialization, a variable template, an alias)
- In the same vein, can this be applied to a partial specialization to prevent subsequent specializations?
- I think the diagnostic should say `note: marked `[[clang::no_specialization]] here`

File: clang/include/clang/Basic/Attr.td (hunk 6 above)
Comment (AaronBallman) [reply clang/include/clang/Basic/Attr.td:3335]: Okay, SGTM!

**Comment by cor3ntin** [comment]
Does it makes sense to also support function templates?
For example, my understanding is that specializing `is_pointer_interconvertible_with_class()` would be UB

**Comment by philnik777** [comment]
> Does it makes sense to also support function templates? For example, my understanding is that specializing `is_pointer_interconvertible_with_class()` would be UB

I can't find anything explicit about function template specializations, so I'm not 100% sure whether it's allowed. Since there is no requirement that the template arguments of any function match the standards-declared ones exactly, I don't think you're allowed to specialize _any_ standard library function.

**Comment by cor3ntin** [comment]
I did think about it more, and I think it would be more consistent to support the attribute on any sort of specializable entity, including functions - regardless of STL needs @AaronBallman @ldionne

**Comment by AaronBallman** [comment]
> I did think about it more, and I think it would be more consistent to support the attribute on any sort of specializable entity, including functions - regardless of STL needs @AaronBallman @ldionne

Agreed; if we're adding a custom attribute, we might as well support it in a general form so users can also make use of it for their needs (which may be different than STL needs).

**Comment by philnik777** [comment]
> > I did think about it more, and I think it would be more consistent to support the attribute on any sort of specializable entity, including functions - regardless of STL needs @AaronBallman @ldionne
> 
> Agreed; if we're adding a custom attribute, we might as well support it in a general form so users can also make use of it for their needs (which may be different than STL needs).

While I don't see much of a use-case for functions, I also don't see much of a downside (assuming implementing the attribute is as simple for functions as it is for variables and classes). Is there anything else that can be specialized?

**Comment by AaronBallman** [comment]
> > > I did think about it more, and I think it would be more consistent to support the attribute on any sort of specializable entity, including functions - regardless of STL needs @AaronBallman @ldionne
> > 
> > 
> > Agreed; if we're adding a custom attribute, we might as well support it in a general form so users can also make use of it for their needs (which may be different than STL needs).
> 
> While I don't see much of a use-case for functions, I also don't see much of a downside (assuming implementing the attribute is as simple for functions as it is for variables and classes). Is there anything else that can be specialized?

Classes, functions, variables
Enumerations... maybe... ish?: https://godbolt.org/z/78qn1hf8d
(https://en.cppreference.com/w/cpp/language/template_specialization)

**Comment by philnik777** [comment]
> > > > I did think about it more, and I think it would be more consistent to support the attribute on any sort of specializable entity, including functions - regardless of STL needs @AaronBallman @ldionne
> > > 
> > > 
> > > Agreed; if we're adding a custom attribute, we might as well support it in a general form so users can also make use of it for their needs (which may be different than STL needs).
> > 
> > 
> > While I don't see much of a use-case for functions, I also don't see much of a downside (assuming implementing the attribute is as simple for functions as it is for variables and classes). Is there anything else that can be specialized?
> 
> Classes, functions, variables Enumerations... maybe... ish?: https://godbolt.org/z/78qn1hf8d (https://en.cppreference.com/w/cpp/language/template_specialization)

That looks horrendous, but I don't think that's specializing the enum. It's specializing the enclosing class and the specialization happens to also define an enum with the same name as the enum in the base class. You can replace the enum with anything you want, including nothing at all.

**Comment by cor3ntin** [comment]
The list is there https://eel.is/c++draft/temp.expl.spec#1

**Comment by philnik777** [comment]
I went for only adding `[[clang::no_specialization]]` to function templates, since the other cases seem even more unlikely to me and seem to have a higher implementation complexity. Is that OK?

**Comment by AaronBallman** [comment]
> I went for only adding `[[clang::no_specialization]]` to function templates, since the other cases seem even more unlikely to me and seem to have a higher implementation complexity. Is that OK?

Your test coverage also handles class templates and variable templates in addition to function templates, which I think covers the most common cases and is fine.

File: clang/include/clang/Basic/AttrDocs.td [hunk 10]
```diff
@@ -981,6 +981,15 @@ Query for this feature with ``__has_attribute(diagnose_if)``.
   }];
 }
 
+def NoSpecializationsDocs : Documentation {
+  let Category = DocCatDecl;
+  let Content = [{
+``[[clang::no_specializations]]`` can be applied to class or variable
```
Comment (erichkeane) [inline clang/include/clang/Basic/AttrDocs.td:987]: You also allow function templates though?

**Comment by philnik777** [comment]
> > I went for only adding `[[clang::no_specialization]]` to function templates, since the other cases seem even more unlikely to me and seem to have a higher implementation complexity. Is that OK?
> 
> Your test coverage also handles class templates and variable templates in addition to function templates, which I think covers the most common cases and is fine.

Ah, yes, I meant "I only added function templates in addition to class and variable templates (which was the original set)".

**Review by erichkeane — COMMENTED** [review]
No real comments, other than FunctionTemplate being left out of the documentation.

File: clang/include/clang/Basic/Attr.td [hunk 11]
```diff
@@ -4637,7 +4649,7 @@ def HLSLResourceBinding: InheritableAttr {
   let AdditionalMembers = [{
   public:
       enum class RegisterType : unsigned { SRV, UAV, CBuffer, Sampler, C, I };
-  
+
```
Comment (erichkeane) [inline clang/include/clang/Basic/Attr.td:4652]: Please don't commit hte unrelated formatting changes.

File: clang/include/clang/Basic/AttrDocs.td [hunk 12]
```diff
@@ -1155,6 +1155,15 @@ Query for this feature with ``__has_attribute(diagnose_if)``.
   }];
 }
 
+def NoSpecializationsDocs : Documentation {
+  let Category = DocCatDecl;
+  let Content = [{
+``[[clang::no_specializations]]`` can be applied to function, class or variable
```
Comment (erichkeane) [inline clang/include/clang/Basic/AttrDocs.td:1161]: Oxford comma?

File: clang/include/clang/Basic/AttrDocs.td [hunk 13]
```diff
@@ -7197,10 +7206,10 @@ alternatives, though the attribute can be used even when the fix can't be automa
   and then use the attribute on the original ``baz()`` to help the users
   update their code to use the new function.
 
-* Attribute attached to fields: The attribute should only be attached to
```
Comment (erichkeane) [inline clang/include/clang/Basic/AttrDocs.td:7200]: Same comments on trailing WS here :/

File: clang/include/clang/Basic/DiagnosticGroups.td [hunk 14]
```diff
@@ -473,6 +473,7 @@ def ExpansionToDefined : DiagGroup<"expansion-to-defined">;
 def FlagEnum : DiagGroup<"flag-enum">;
 def IncrementBool : DiagGroup<"increment-bool", [DeprecatedIncrementBool]>;
 def InfiniteRecursion : DiagGroup<"infinite-recursion">;
+def InvalidSpecialization : DiagGroup<"invalid-specialization">;
```
Comment (erichkeane) [inline clang/include/clang/Basic/DiagnosticGroups.td:476]: Only 1 warning here, no reason to make a non-inline group.

**Review by erichkeane — APPROVED** [review]
Nits only, otherwise LGTM.

File: clang/include/clang/Basic/DiagnosticGroups.td (hunk 14 above)
Comment (philnik777) [reply clang/include/clang/Basic/DiagnosticGroups.td:476]: What do you mean?

Comment (philnik777) [reply clang/include/clang/Basic/DiagnosticGroups.td:476]: Ah, I didn't realize you could do that.

**Comment by philnik777** [comment]
@AaronBallman @cor3ntin are you happy as well?

**Review by AaronBallman — APPROVED** [review]
LGTM!

File: clang/lib/Sema/SemaDeclAttr.cpp [hunk 15]
```diff
@@ -1212,6 +1212,14 @@ static void handlePreferredName(Sema &S, Decl *D, const ParsedAttr &AL) {
         << TT->getDecl();
 }
 
+static void handleNoSpecializations(Sema &S, Decl *D, const ParsedAttr &AL) {
+  StringRef Message;
+  if (AL.getNumArgs() != 0)
+    S.checkStringLiteralArgumentAttr(AL, 0, Message);
```
Comment (shafik) [inline clang/lib/Sema/SemaDeclAttr.cpp:1218]: static analysis flags this b/c you are not checking the return value of `checkStringLiteralArgumentAttr` and AFAICT every other use is checking the value.

Please confirm the intent and please fix to check the return value or please modify to purposely ignore it.

**Review by shafik — COMMENTED** [review]
Flagging what looks like a bug

File: clang/lib/Sema/SemaDeclAttr.cpp (hunk 15 above)
Comment (philnik777) [reply clang/lib/Sema/SemaDeclAttr.cpp:1218]: I'm not 100% sure what the correct thing to do is. I don't think there's harm in adding the attribute without a string if the arguments are incorrect, since they don't affect what the attribute does in a significant way. OTOH there isn't that much benefit in adding it, since the program is ill-formed anyways. @AaronBallman any thoughts on this?

P.S. why isn't `checkStringLiteralArgumentAttr` marked `[[nodiscard]]`?

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:1218]: >P.S. why isn't checkStringLiteralArgumentAttr marked [[nodiscard]]?

Pretty sure this function, or its origin predate `[[nodiscard]]` by a decade or so :)  Also, the use you have is completely reasonable here.  The `Message` argument is left zero-initalized, which is the behavior you want.

I don't see any good reason to change the code here, I think the SA tool isn't really adding value here, and I don't see anything I'd change.

Comment (shafik) [reply clang/lib/Sema/SemaDeclAttr.cpp:1218]: It stands out as completely inconsistent with the rest of the file, literally everywhere else it is checked and they return early. I am guessing if we return early here we should just get the proper diagnostic, right? Is this case somehow different from all the two dozen or so other uses?

Comment (erichkeane) [reply clang/lib/Sema/SemaDeclAttr.cpp:1218]: Yes, it is inconsistent, but not really an issue. About 1/2 the uses have some additional processing that needs to happen, so the early return is sensible, though perhaps they could better reflect the code in that case.

But the other 1/2 could very well have their return removed with no negative effects (and likely, an improvement, as the error-AST would better reflect the syntax).

The only reason NOT to do a mass-changeover is that we're not sure which were written in a way to not properly expect an empty string here in codegen/etc.

That said, a bug report marked 'good-first-issue' for someone to go through the uses and ensure we are tolerant of them would be a good idea.  @shafik: mind filing one?
